include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -294,6 +294,37 @@ def ConstantOp : MooreOp<"constant", [Pure]> {
   ];
 }
 
+def Parameter : I32EnumAttrCase<"Parameter", 0, "parameter">;
+def LocalParameter : I32EnumAttrCase<"LocalParameter", 1, "localparam">;
+def SpecParameter : I32EnumAttrCase<"SpecParameter", 2, "specparam">;
+
+def NamedConstAttr : I32EnumAttr<"NamedConst", "elaboration-time constants", 
+       [Parameter, LocalParameter, SpecParameter]>{
+    let cppNamespace = "circt::moore";
+}
+
+def NamedConstantOp : MooreOp<"named_constant", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
```

Comment:
If this op is approved. Maybe we should add `ConstantLike` and `Pure` to its trait list.

Reply:
Cool! It's great to have an op to materialize the compile-time constants and preserve them. I don't think you can use `ConstantLike` here since the op takes a single operand, unlike `ConstantOp` which has no operands and just a single constant value attribute. `Pure` could work, but that would mean that if the `named_constant` result is unused (which I think it will be in a lot of cases, for example if the constant value was only used to parametrize types) then CSE/dead code elimination will delete the `named_constant` op again. That might not be what we want.

Reply:
I get it! Looks like I misunderstod.